staging/src/k8s.io/component-base/featuregate/testing/feature_gate.go

```diff
@@ -18,18 +18,49 @@ package testing
 
 import (
 	"fmt"
-	"testing"
+	"sync"
 
 	"k8s.io/component-base/featuregate"
 )
 
-// SetFeatureGateDuringTest sets the specified gate to the specified value, and returns a function that restores the original value.
-// Failures to set or restore cause the test to fail.
+var (
+	overwrittenFeatureFlags map[featuregate.Feature]struct{}
+	lock                    sync.Mutex
+)
+
+func init() {
+	overwrittenFeatureFlags = map[featuregate.Feature]struct{}{}
+}
+
+// SetFeatureGateDuringTest sets the specified gate to the specified value for duration of the test.
+// Fails when it detects second call to the same flag or is unable to set or restore feature flag.
+// Returns empty cleanup function to maintain the old function signature that uses defer.
+// TODO: Remove defer from calls to SetFeatureGateDuringTest and update hack/verify-test-featuregates.sh when we can do large scale code change.
+//
+// WARNING: Can leak set variable when called in test calling t.Parallel(), however second attempt to set the same feature flag will cause fatal.
 //
 // Example use:
 //
 // defer featuregatetesting.SetFeatureGateDuringTest(t, utilfeature.DefaultFeatureGate, features.<FeatureName>, true)()
-func SetFeatureGateDuringTest(tb testing.TB, gate featuregate.FeatureGate, f featuregate.Feature, value bool) func() {
+func SetFeatureGateDuringTest(tb testingT, gate featuregate.FeatureGate, f featuregate.Feature, value bool) func() {
+	lock.Lock()
```

Comment:
if you lock during the entire test duration what is the benefit of t.Parallel?

Reply:
We don't lock during test duration, locking is just to protect critical section.